### PR TITLE
Move repository model into models/repo

### PR DESCRIPTION
**Pull request by lunny**
This is the first step to move repository related beans. To make it review easier, I will put others in future PRs.

### CODE REVIEW

**Comment by wxiaoguang** [comment]
It would be better to keep model related prefix for functions in root `models` package.

For example, `repoCheckUnitUser` instead of `checkUnitUser`, `repoReconsiderIssueAssignees` instead of `reconsiderIssueAssignees`, `RepoDeleteCollaboration` instead of `DeleteCollaboration`, `RepoAddDeletedBranch` ... etc


And `JSONUnmarshalHandleDoubleEncode` should be put somewhere else, `auth/source` shouldn't depend on `repo`


Otherwise, LGTM

**Comment by lunny** [comment]
> AddDeletedBranch

I changed most of them but kept `DeleteCollaboration`, `AddDeletedBranch` because I think the repo prefix is not a must word.

**Review by wxiaoguang — APPROVED** [review]

File: models/branches.go [hunk 1]
```diff
@@ -393,20 +394,20 @@ func UpdateProtectBranch(repo *Repository, protectBranch *ProtectedBranch, opts
...
 	protectedBranches := make([]*ProtectedBranch, 0)
-	return protectedBranches, db.GetEngine(db.DefaultContext).Find(&protectedBranches, &ProtectedBranch{RepoID: repo.ID})
+	return protectedBranches, db.GetEngine(db.DefaultContext).Find(&protectedBranches, &ProtectedBranch{RepoID: repoID})
 }
 
 // GetBranchProtection get the branch protection of a branch
-func (repo *Repository) GetBranchProtection(branchName string) (*ProtectedBranch, error) {
-	return GetProtectedBranchBy(repo.ID, branchName)
+func GetBranchProtection(repoID int64, branchName string) (*ProtectedBranch, error) {
+	return GetProtectedBranchBy(repoID, branchName)
```
Comment (KN4CK3R) [inline models/branches.go]: Could replace `GetBranchProtection` with `GetProtectedBranchBy`.

**Review by KN4CK3R — APPROVED** [review]

File: models/branches.go (hunk 1 above)
Comment (lunny) [reply models/branches.go]: Good catch. Done.

File: modules/context/repo.go [hunk 2]
```diff
@@ -361,21 +364,24 @@ func repoAssignment(ctx *Context, repo *models.Repository) {
 
 	if repo.IsMirror {
 		var err error
-		ctx.Repo.Mirror, err = models.GetMirrorByRepoID(repo.ID)
+		mirror, err := repo_model.GetMirrorByRepoID(repo.ID)
```
Comment (zeripath) [inline modules/context/repo.go:367]: This causes #18204

**Pull request merged**